### PR TITLE
Add support for the Friendly Forge Format (F3)

### PR DESCRIPTION
**Pull request by earl-warren**
This pull request is a mirror of a corresponding pull request on gitea.com, please [go there to discuss and comment](https://gitea.com/earl-warren/gitea/pulls/2).

depends on:
 - [x] https://lab.forgefriends.org/friendlyforgeformat/gof3/-/issues/21

---

Continuation of [this previous pull request](https://github.com/go-gitea/gitea/pull/20122)

### CODE REVIEW

**Comment by earl-warren** [comment]
How can it be that [lint on windows reports an error](https://drone.gitea.io/go-gitea/gitea/60169/1/6) and [lint on GNU/Linux does not](https://drone.gitea.io/go-gitea/gitea/60169/1/5)?

**Comment by silverwind** [comment]
> How can it be that [lint on windows reports an error](https://drone.gitea.io/go-gitea/gitea/60169/1/6) and [lint on GNU/Linux does not](https://drone.gitea.io/go-gitea/gitea/60169/1/5)?

I think the only logical reason would be presence of `//go:build` tags that run only on Windows.

**Comment by earl-warren** [comment]
> > How can it be that [lint on windows reports an error](https://drone.gitea.io/go-gitea/gitea/60169/1/6) and [lint on GNU/Linux does not](https://drone.gitea.io/go-gitea/gitea/60169/1/5)?
> 
> I think the only logical reason would be presence of `//go:build` tags that run only on Windows, but that doesn't seem to be the case here. Might just restart and see if it goes away.

But there are no `//go:build tags` in this PR. Is there a way to re-launch the CI and verify this is not a false negative?

**Comment by earl-warren** [comment]
Now that's a [new and interesting false negative](https://drone.gitea.io/go-gitea/gitea/60247).

image

File: services/f3/driver/driver.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+	var ok bool
+	o.options, ok = options.(*Options)
+	if !ok {
+		panic(fmt.Errorf("unexpected type %T", options))
+	}
+}
+
+func (o *Gitea) Init(options gof3.OptionsInterface) {
+	o.SetOptions(options)
+	o.perPage = 100
```
Comment (6543) [inline services/f3/driver/driver.go]: That's a instance config and shuld not be hardcoded

Comment (earl-warren) [reply services/f3/driver/driver.go]: fixed

**Comment by earl-warren** [comment]
The CI is broken.

image

**Comment by delvh** [comment]
We know. https://github.com/go-gitea/gitea/pull/21108

**Comment by earl-warren** [comment]
Now that #21108 is merged, the CI is back to another false negative: [failing lint on windows](https://drone.gitea.io/go-gitea/gitea/60315) but not on GNU/Linux, despite the absence of build tags. 

image

**Comment by earl-warren** [comment]
> Now that #21108 is merged, the CI is back to another false negative: [failing lint on windows](https://drone.gitea.io/go-gitea/gitea/60315) but not on GNU/Linux, despite the absence of build tags.
> 
> image

This pull request is not compiled for windows to workaround the linter false negative.

**Comment by earl-warren** [comment]
The arm64 test image was rebuilt with 1.19.1, trying again.

**Comment by silverwind** [comment]
Side note:

> This pull request is a mirror of a corresponding pull request on gitea.com, please [go there to discuss and comment](https://gitea.com/earl-warren/gitea/pulls/2).

gitea core development takes place on GitHub, so it should be reviewed/discussed here :)

File: services/f3/driver/pull_request.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+	for _, label := range pullRequest.Labels {
+		labels = append(labels, &issues_model.Label{Name: label})
+	}
+
+	// assignees := make([]*user_model.User, 0, len(pullRequest.Assignees))
+	// for _, a := range pullRequest.Assignees {
+	// 	assignees = append(assignees, &user_model.User{
+	// 		Name: a,
+	// 	})
+	// }
```
Comment (silverwind) [inline services/f3/driver/pull_request.go]: No commented-out code please. Ignore if this is still part of the WIP :)

Comment (earl-warren) [reply services/f3/driver/pull_request.go]: Good catch, it was an oversight.

**Comment by earl-warren** [comment]
> Side note:
> 
> > This pull request is a mirror of a corresponding pull request on gitea.com, please [go there to discuss and comment](https://gitea.com/earl-warren/gitea/pulls/2).
> 
> gitea core development takes place on GitHub, so it should be reviewed/discussed here :)

I did not want to force people to agree to the GitHub terms of service to participate in this particular pull request. Moving the discussion to gitea.com is a way to be more inclusive in a way that is not disruptive to the Gitea development process.

**Comment by earl-warren** [comment]
> > Side note:
> > > This pull request is a mirror of a corresponding pull request on gitea.com, please [go there to discuss and comment](https://gitea.com/earl-warren/gitea/pulls/2).
> > 
> > 
> > gitea core development takes place on GitHub, so it should be reviewed/discussed here :)
> 
> I did not want to force people to agree to the GitHub terms of service to participate in this particular pull request. Moving the discussion to gitea.com is more inclusive in a way that is not disruptive to the Gitea development process.

I commented further on the [issue you opened](https://github.com/go-gitea/gitea/issues/21247) that is related to this topic.

**Comment by Gusted** [comment]
Just to mention this again, a valid point was raised on the [previous PR for F3](https://github.com/go-gitea/gitea/pull/20122#issuecomment-1193073570). I would like to state again that Loïc and the other contributors will change the license to MIT when this PR is approved to be merged.

**Comment by 6543** [comment]
@earl-warren could you add me to your fork as contrib? so i can update that pull ... also if you would not force-push I would aprechiate!

**Comment by 6543** [comment]
blocked by https://lab.forgefriends.org/friendlyforgeformat/gof3/-/issues/21

File: modules/setting/f3.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+var (
+	F3 = struct {
+		Enabled bool
+	}{
+		Enabled: true,
+	}
+)
+
+func newF3Service() {
+	if err := Cfg.Section("F3").MapTo(&F3); err != nil {
```
Comment (6543) [inline modules/setting/f3.go:21]: all other sections are lowercase
```suggestion
	if err := Cfg.Section("f3").MapTo(&F3); err != nil {
```

File: custom/conf/app.example.ini [hunk 4]
```diff
@@ -2294,6 +2294,15 @@ ROUTER = console
 ;; If a domain is allowed by ALLOWED_DOMAINS, this option will be ignored.
 ;ALLOW_LOCALNETWORKS = false
 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;[F3]
```
Comment (6543) [inline custom/conf/app.example.ini:2299]:
```suggestion
;[f3]
```

**Comment by 6543** [comment]
please update branch & update f3 dep

**Pull request closed without merging**